File: python/creational/factory-method.py
from abc import ABC, abstractmethod

# Product interface
class Notification(ABC):

    @abstractmethod
    def send(self):
        pass

# Product implementation
class SMSNotification(Notification):
    def send(self, message):
        print(f"Send SMS: {message}")

class EmailNotitication(Notification):
    def send(self, message):
        print(f"Send Email: {message}")

# Creator interface
class NotificationFactory(ABC):
    @abstractmethod
    def create_notification(self):
        pass

# Concerte creators
class SMSNotificationFactory(NotificationFactory):
    def create_notification(self) -> Notification:
        return SMSNotification()
    
class EmailNotificationFactory(NotificationFactory):
    def create_notification(self) -> Notification:
        return EmailNotitication()
    

def notify(factory: NotificationFactory, message: str):
    notification = factory.create_notification()
    notification.send(message)
if __name__ == "__main__":
    
    factories = {
        "email": EmailNotificationFactory(),
        "sms": SMSNotificationFactory(),
    }

  
    notify(factories["sms"], "Hello ...")
   

File: go/creational/factory-method.go
package main

import "fmt"

// Product interface
type Notification interface {
	send(message string)
}

// Concerete products
type SMSNotification struct{}

func (n *SMSNotification) send(message string) {
	fmt.Println("Send SMS: ", message)
}

type EmailNotification struct{}

func (n *EmailNotification) send(message string) {
	fmt.Println("Send Email: ", message)
}

// Creator Interface
type NotificationFactory interface {
	CreateNotification() Notification
}

// Concerete Creators

type EmailNotificationFactory struct{}

func (n *EmailNotificationFactory) CreateNotification() Notification {
	return &EmailNotification{}
}

type SMSNotificationFactory struct{}

func (n *SMSNotificationFactory) CreateNotification() Notification {
	return &SMSNotification{}
}

func notify(factory NotificationFactory, message string) {
	notificationFactory := factory.CreateNotification()
	notificationFactory.send(message)
}

func main() {

	var factory NotificationFactory
	notificationType := "email"

	switch notificationType {
	case "email":
		factory = &EmailNotificationFactory{}
	case "sms":
		factory = &SMSNotificationFactory{}
	default:
		fmt.Println("Invalid notification type")
		return
	}

	notify(factory, "Hello..")
}
